components/sup/src/event.rs

```diff
@@ -109,9 +97,49 @@ pub struct EventCore {
     pub supervisor_id: String,
 }
 
+/// Send an event for the start of a Service.
+pub fn service_started(service: &Service) {
+    if stream_initialized() {
+        publish(ServiceStartedEvent { service_metadata:    Some(service.to_service_metadata()),
+                                      supervisor_metadata: None, });
+    }
+}
+
+/// Send an event for the stop of a Service.
+pub fn service_stopped(service: &Service) {
+    if stream_initialized() {
+        publish(ServiceStoppedEvent { service_metadata:    Some(service.to_service_metadata()),
+                                      supervisor_metadata: None, });
+    }
+}
+
+////////////////////////////////////////////////////////////////////////
+
+/// Internal helper function to know whether or not to go to the trouble of
+/// creating event structures. If the event stream hasn't been
+/// initialized, then we shouldn't need to do anything.
+fn stream_initialized() -> bool { EVENT_STREAM.try_get::<EventStream>().is_some() }
+
+/// Publish an event. This is the main interface that client code will
+/// use.
+///
+/// If `init_stream` has not been called already, this function will
+/// be a no-op.
+fn publish(mut event: impl EventMessage) {
+    // TODO: incorporate the current timestamp into the rendered event
+    // (which will require tweaks to the rendering logic, but we know
+    // that'll need to be updated anyway).
+    if let Some(e) = EVENT_STREAM.try_get::<EventStream>() {
+        event.supervisor_metadata(EVENT_CORE.get::<EventCore>().to_supervisor_metadata());
+        if let Ok(bytes) = event.to_bytes() {
```

Comment:
Why would event serialization ever fail?

Reply:
I think it'd be pretty weird to fail, but this is basically just dealing with the fact that the underlying Prost code returns a `Result`.

I can dig in a bit further to see if we can safely push this down into the `event.to_bytes()` method.

Reply:
@baumanj Indeed, we can push it down: https://github.com/habitat-sh/habitat/pull/6344/commits/0a2c9fb54a02a2bc70cff0e3e0cf32831ff56176